woocommerce/woocommerce-gutenberg-products-block
# Support updating quantity of cart items & sold_individually product option

Fixes #1746 #1822

Adds support for changing cart items by hooking up the cart line item `QuantitySelector` to a new action for updating quantity server side via API. As part of this, the `QuantitySelector` no longer tracks its value (`quantity`) using internal state – this must be managed outside the component and fed in via props (and changes via `onChange` prop) - aka a [controlled component](https://reactjs.org/docs/forms.html#controlled-components). Storybook has been updated to the new usage.

Update: this PR also includes support for a min and max value in Quantity Selector, to avoid issues when making API requests to set cart item quantity to zero (item disappears during request handling). This means that the shopper has to click "remove item" to remove a cart item; this can't be done by decrementing quantity down to 0. 

The fix for #1822 has been rolled into this PR, as it is closely related. If a product has `sold_individually` option set, then the quantity picker is limited to maximum of 1. A new field has been added to `cart/items` api endpoint for `sold_individually`.

Note: in future we might consider allowing quantity => 0 as a way to remove an item, perhaps when we update the cart/items API to always return complete cart API response. However in this PR I've locked this behaviour down so it's stable.

### Screenshots

- **Album** product is `sold_individually`.

<img width="898" alt="Screen Shot 2020-02-27 at 1 56 01 PM" src="https://user-images.githubusercontent.com/4167300/75402108-e4e67200-5968-11ea-839c-bbbdd508bae9.png">

In cart:

- Notice quantity picker has disabled add/subtract buttons as needed.
- **Album** has max quantity of 1, because `sold_individually`=true.
- Other products have min quantity of 1, to prevent 500 errors when POSTing `quantity`=0.

<img width="821" alt="Screen Shot 2020-03-02 at 2 55 27 PM" src="https://user-images.githubusercontent.com/4167300/75639479-df08cd80-5c95-11ea-8f5f-b64845a4709a.png">


### How to test the changes in this Pull Request:

1. Add a cart block to a page and publish.
2. Add some items to the cart and view cart page.
23. Edit products and enable `sold_individually` option.
3. Change the quantity of items and remove items (etc). Ensure the cart block updates correctly and the server-side cart object is updated (i.e. compare with traditional cart shortcode).
